doc_source/external-snat.md

```diff
@@ -1,6 +1,6 @@
 # External Source Network Address Translation \(SNAT\)<a name="external-snat"></a>
 
-By default, the [Amazon VPC CNI plugin for Kubernetes](https://github.com/aws/amazon-vpc-cni-k8s) configures pods with source network address translation \(SNAT\) enabled\. This sets the return address for a packet to the primary public IP of the instance and allows for communication with the internet\. In this default configuration, when you use an internet gateway and a public address, the return packet is routed to the correct Amazon EC2 instance\.
```

Comment:
I definitely agree that covering the case of public and the case of private subnets distinctly can make the documentation clearer. I have a different understanding of what you explained above, I'll try to quickly explain my version.

Let me first cover the case for worker nodes in a public subnet. I believe that an EC2 instance does not know its own public IP. I base this on the following explanation, taken from the official VPC Internet Gateway documentation (https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html):

> Your instance is only aware of the private (internal) IP address space defined within the VPC and subnet. The internet gateway logically provides the one-to-one NAT on behalf of your instance, so that when traffic leaves your VPC subnet and goes to the internet, the reply address field is set to the public IPv4 address or Elastic IP address of your instance, and not its private IP address. Conversely, traffic that's destined for the public IPv4 address or Elastic IP address of your instance has its destination address translated into the instance's private IPv4 address before the traffic is delivered to the VPC.

If this assumption holds, then the CNI plugin cannot do a translation from secondary private to public, as it doesn't even know the public address. What it can do is translate from the secondary private address to the primary private address. This is a must, as the Internet Gateway can only do the mapping between the primary private and the public address, it cannot map between secondary private addresses and the public IP. This is exactly the reason, why the external SNAT cannot be enabled for public subnets.

The case of nodes in a private subnet is different as the NAT Gateway can route not only to the primary private IP but to the secondary private IPs as well. So it is possible to not do SNAT on the worker node itself, the NAT Gateway will take care of it.

Still, the SNAT that can take place on the node is always between secondary private IP and primary private IP, the public IP is never involved there.

I think the explainer image should be also updated (maybe even have one for the public and one for the private case). Do you know if there is a supported way to update images in PRs?

Reply:
Any thoughts on the above?

Reply:
@tkornai my apologies. still mulling this over. want to chat with @anguslees about this further. I don't believe there is a way of updating the images in a PR through Github. You'd need to `git commit` a new image.

Reply:
Thanks, @jaypipes for getting back to me. I'm happy to propose updated images in case I can get access to the source of the originals and if you can point me to the diagramming tool used.

Reply:
Since networking within a cluster is "flat" we don't SNAT any traffic within the VPC, but we SNAT all external traffic to come from the node host IP. Starting with v1.6.0, we will add [support to disable SNAT inside peered VPCs](https://github.com/aws/amazon-vpc-cni-k8s/pull/520) as well through the new `AWS_VPC_K8S_CNI_EXCLUDE_SNAT_CIDRS ` environment variable. 

I could help update the image as well.

Reply:
@mogren, how about this text:

"By default, the [Amazon VPC CNI plugin for Kubernetes](https://github.com/aws/amazon-vpc-cni-k8s) configures pods with source network address translation (SNAT) enabled for traffic that leaves the VPC. Communication within the VPC (such as pod to pod) is direct and SNAT does not occur.

For pods running on worker nodes that are in a VPC with a private subnet, SNAT will set the return address for a packet to the IP of the subnet's associated Internet Gateway. For pods running on worker nodes that are in a VPC with a public subnet, SNAT will set the return address for a packet to the worker node's public IP address. This allows for communication from the pod to the Internet and ensures the return packet is routed to the correct Amazon EC2 instance."

Reply:
@nrdlngr LGTM!

Reply:
@nrdlngr My understanding is that private subnets are always accessing the internet through NAT gateways and never through Internet Gateways. I base this on the official documentation quoted below.

> An internet gateway serves two purposes: to provide a target in your VPC route tables for internet-routable traffic, and to perform network address translation (NAT) for instances that have been assigned **public** IPv4 addresses.

source: https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html

> You can use a network address translation (NAT) gateway to enable instances in a **private subnet** to connect to the internet or other AWS services, but prevent the internet from initiating a connection with those instances. For more information about NAT, see NAT.

source: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html

I also quoted the official documentation in my previous reply about how EC2 instances are not aware of their public IP - so SNAT on the node can only happen between private IPs. That is secondary private IPs get changed to the primary private IP of the node.

Reply:
It is true that instances in private subnets use NAT gateways for communication to the internet, but that must also go through an Internet Gateway. Without and Internet Gateway, nothing in your VPC can connect to the internet.

Also, the CNI uses the EC2 DescribeInstances API call to get information about the instance it is running on, so it doesn't matter what the host knows or doesn't know about its IP addresses. It gets permissions for this call from the worker node's instance IAM role.

Reply:
Hi @nrdlngr, sorry for being so persistent about this issue, but I think the information in the documentation is still wrong. I believe your two statements above are mostly true, but they do not address my primary concern.

My understanding is that inside the VPC all routing is based on private IPs. Translation from private IP to public IP only happens on the InternetGateway. This is confirmed by the official documentation:

> To enable communication over the internet for IPv4, your instance must have a public IPv4 address or an Elastic IP address that's associated with a private IPv4 address on your instance. Your instance is only aware of the private (internal) IP address space defined within the VPC and subnet. The internet gateway logically provides the one-to-one NAT on behalf of your instance, so that when traffic leaves your VPC subnet and goes to the internet, the reply address field is set to the public IPv4 address or Elastic IP address of your instance, and not its private IP address. Conversely, traffic that's destined for the public IPv4 address or Elastic IP address of your instance has its destination address translated into the instance's private IPv4 address before the traffic is delivered to the VPC.

Hence the VPC CNI plugin should never set the source IP to a public IP.

If my understanding is incorrect, please help me understand where I'm mistaken.